lib/wlib/memory/Memory.cpp

```diff
@@ -47,24 +60,96 @@ T nextHigher(T k) {
     return k + 1;
 }
 
-void memory_init() {}
+/**
+ * Sometimes we want to have restrictions on block sizes and hence this
+ * function can be used to specify restrictions
+ *
+ * @param blockSize size of the block provided
+ * @return the size of block after restrictions
+ */
+constexpr inline uint16_t applyBlockSizeRestrictions(uint32_t blockSize) {
+    if (blockSize > 256 && blockSize <= 512)
+        blockSize = 300;
+    else if (blockSize > 512 && blockSize <= 1024)
+        blockSize = 400;
+
+    return blockSize;
+}
+
+/**
+ * Template struct to create allocators. This recursively creates allocators and hence makes
+ * a memory pool. This is the recursive step of the process
+ * @tparam powStart starting power of 2 for memory of Allocator
+ * @tparam from number to start from
+ * @tparam to number to end at
+ */
+template<uint16_t powStart, uint16_t from, uint16_t to>
+struct insert {

```

Comment:
This tmp makes me w3t

Reply:
most complicated template I ever wrote :p